File: Algo/Leetcode/099RecoverBinarySearchTree.cpp
//1. Use inorder sort for binary search tree. Change the wrong pointers.
class Solution {
public:
    void recoverTree(TreeNode* root) {
        vector<TreeNode *> res = inorderTraversal(root);
        
        TreeNode * l = nullptr;
        TreeNode * r = nullptr;
        for(int i=0;i<res.size()-1;i++){
            if(res[i]->val > res[i+1]->val) {l = res[i];break;}
        }
        for(int i=res.size()-1;i>0;i--){
            if(res[i]->val < res[i-1]->val) {r = res[i];break;}
        }
        int temp = r->val;
        r->val = l->val;
        l->val = temp;
    }
    
    vector<TreeNode *> inorderTraversal(TreeNode* root) {
        vector<TreeNode *> res;
        if(!root) return res;
        
        inorderTraversal(root,res);
        return res;
    }
    
    void inorderTraversal(TreeNode* root, vector<TreeNode *>& res) {
        if(!root) return;
        
        if(root->left) inorderTraversal(root->left,res);
        res.push_back(root);
        if(root->right) inorderTraversal(root->right,res);    
    }
};

File: Algo/Leetcode/099RecoverBinarySearchTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):

    def __init__(self):
        self.res = []

    def recoverTree(self, root):
        """
        :type root: TreeNode
        :rtype: None Do not return anything, modify root in-place instead.
        """
        if root is None:
            return

        self.dfs(root)

        n = len(self.res)
        i = 0
        while i < n and self.res[i].val < self.res[i + 1].val:
            i += 1
        j = n - 1
        while j > 0 and self.res[j].val > self.res[j - 1].val:
            j -= 1

        tmp = self.res[i].val
        self.res[i].val = self.res[j].val
        self.res[j].val = tmp

    def dfs(self, root):
        if root is None:
            return

        self.dfs(root.left)
        self.res.append(root)  # push the node
        self.dfs(root.right)